File: NextApp/Android/core/data/src/main/kotlin/jp/komehyappyo/member/next/core/data/DiaryBackupService.kt
package jp.komehyappyo.member.next.core.data

import android.util.Base64
import jp.komehyappyo.member.next.core.model.Diary
import jp.komehyappyo.member.next.core.model.DiaryMood
import kotlinx.coroutines.flow.first
import org.json.JSONArray
import org.json.JSONObject
import java.security.MessageDigest
import java.time.Instant
import java.util.UUID

class DiaryBackupService(
    private val repository: DiaryRepository,
    private val photoStore: DiaryPhotoStore,
) {
    suspend fun exportData(now: Instant = Instant.now()): ByteArray {
        val diaries = repository.observeAll().first()
        val diaryArray = JSONArray()
        diaries.forEach { diary ->
            val photos = JSONArray()
            diary.photoUrls.forEach { reference ->
                val data = photoStore.load(reference)
                photos.put(
                    JSONObject()
                        .put("dataBase64", Base64.encodeToString(data, Base64.NO_WRAP))
                        .put("sha256", sha256(data)),
                )
            }
            diaryArray.put(
                JSONObject()
                    .put("id", diary.id.toString())
                    .put("userId", diary.userId)
                    .put("title", diary.title)
                    .put("body", diary.body)
                    .put("mood", diary.mood.toBackupValue())
                    .put("createdAtEpochMillis", diary.createdAt.toEpochMilli())
                    .put("updatedAtEpochMillis", diary.updatedAt.toEpochMilli())
                    .put("photos", photos),
            )
        }
        return JSONObject()
            .put("format", FORMAT_IDENTIFIER)
            .put("version", CURRENT_VERSION)
            .put("exportedAtEpochMillis", now.toEpochMilli())
            .put("diaries", diaryArray)
            .toString(2)
            .toByteArray(Charsets.UTF_8)
    }

    suspend fun importData(data: ByteArray): Int {
        val root = runCatching { JSONObject(data.toString(Charsets.UTF_8)) }
            .getOrElse { throw DiaryBackupException.InvalidFormat }
        if (root.optString("format") != FORMAT_IDENTIFIER) {
            throw DiaryBackupException.InvalidFormat
        }
        if (root.optInt("version") != CURRENT_VERSION) {
            throw DiaryBackupException.UnsupportedVersion
        }
        val entries = root.optJSONArray("diaries") ?: throw DiaryBackupException.InvalidFormat
        val existing = repository.observeAll().first().associateBy(Diary::id)
        var restoredCount = 0
        for (entryIndex in 0 until entries.length()) {
            val entry = entries.optJSONObject(entryIndex)
                ?: throw DiaryBackupException.InvalidFormat
            val id = runCatching { UUID.fromString(entry.getString("id")) }
                .getOrElse { throw DiaryBackupException.InvalidFormat }
            val mood = diaryMoodFromBackupValue(entry.optString("mood"))
                ?: throw DiaryBackupException.InvalidFormat
            val photos = entry.optJSONArray("photos") ?: JSONArray()
            if (photos.length() > Diary.MAXIMUM_PHOTO_COUNT) {
                throw DiaryBackupException.InvalidFormat
            }

            val decodedPhotos = (0 until photos.length()).map { photoIndex ->
                val photo = photos.optJSONObject(photoIndex)
                    ?: throw DiaryBackupException.InvalidPhoto
                val bytes = runCatching {
                    Base64.decode(photo.getString("dataBase64"), Base64.DEFAULT)
                }.getOrElse { throw DiaryBackupException.InvalidPhoto }
                if (
                    bytes.isEmpty() ||
                    bytes.size > LocalDiaryPhotoStore.MAXIMUM_PHOTO_BYTES ||
                    sha256(bytes) != photo.optString("sha256")
                ) {
                    throw DiaryBackupException.InvalidPhoto
                }
                bytes
            }

            val newReferences = mutableListOf<String>()
            try {
                decodedPhotos.forEach {
                    newReferences += photoStore.saveJpeg(it, id)
                }
                val diary = Diary(
                    id = id,
                    userId = entry.optString("userId", "guest"),
                    title = entry.getString("title"),
                    body = entry.optString("body"),
                    mood = mood,
                    photoUrls = newReferences,
                    createdAt = Instant.ofEpochMilli(entry.getLong("createdAtEpochMillis")),
                    updatedAt = Instant.ofEpochMilli(entry.getLong("updatedAtEpochMillis")),
                )
                repository.save(diary)
                existing[id]?.photoUrls?.forEach { oldReference ->
                    runCatching { photoStore.delete(oldReference) }
                }
                restoredCount += 1
            } catch (error: Throwable) {
                newReferences.forEach { runCatching { photoStore.delete(it) } }
                throw error
            }
        }
        return restoredCount
    }

    private fun sha256(data: ByteArray): String =
        MessageDigest.getInstance("SHA-256")
            .digest(data)
            .joinToString("") { "%02x".format(it) }

    private fun DiaryMood.toBackupValue(): String = when (this) {
        DiaryMood.VeryGood -> "veryGood"
        DiaryMood.Good -> "good"
        DiaryMood.Neutral -> "neutral"
        DiaryMood.SlightlyBad -> "slightlyBad"
        DiaryMood.Bad -> "bad"
    }

    private fun diaryMoodFromBackupValue(value: String): DiaryMood? = when (value) {
        "veryGood" -> DiaryMood.VeryGood
        "good" -> DiaryMood.Good
        "neutral" -> DiaryMood.Neutral
        "slightlyBad" -> DiaryMood.SlightlyBad
        "bad" -> DiaryMood.Bad
        else -> null
    }

    companion object {
        const val FORMAT_IDENTIFIER = "org-portal-diary-backup"
        const val CURRENT_VERSION = 1
    }
}

sealed class DiaryBackupException : Exception() {
    data object InvalidFormat : DiaryBackupException()
    data object UnsupportedVersion : DiaryBackupException()
    data object InvalidPhoto : DiaryBackupException()
}

File: NextApp/iOS/OrgPortalNext/Core/DataLayer/DiaryBackupService.swift
import CryptoKit
import Foundation
import Model

public enum DiaryBackupError: Error, Equatable {
    case invalidFormat
    case unsupportedVersion
    case invalidPhoto
}

private struct DiaryBackupEnvelope: Codable {
    let format: String
    let version: Int
    let exportedAtEpochMillis: Int64
    let diaries: [DiaryBackupEntry]
}

private struct DiaryBackupEntry: Codable {
    let id: String
    let userId: String
    let title: String
    let body: String
    let mood: String
    let createdAtEpochMillis: Int64
    let updatedAtEpochMillis: Int64
    let photos: [DiaryBackupPhoto]
}

private struct DiaryBackupPhoto: Codable {
    let dataBase64: String
    let sha256: String
}

@MainActor
public final class DiaryBackupService {
    public static let formatIdentifier = "org-portal-diary-backup"
    public static let currentVersion = 1

    private let repository: DiaryRepository
    private let photoStore: DiaryPhotoStoring

    public init(repository: DiaryRepository, photoStore: DiaryPhotoStoring) {
        self.repository = repository
        self.photoStore = photoStore
    }

    public func exportData(now: Date = .now) async throws -> Data {
        let diaries = try await repository.fetchAll()
        let entries = try diaries.map { diary in
            let photos = try diary.photoUrls.map { reference in
                let data = try photoStore.loadData(reference: reference)
                return DiaryBackupPhoto(
                    dataBase64: data.base64EncodedString(),
                    sha256: Self.sha256(data)
                )
            }
            return DiaryBackupEntry(
                id: diary.id.uuidString,
                userId: diary.userId,
                title: diary.title,
                body: diary.body,
                mood: diary.mood.rawValue,
                createdAtEpochMillis: Self.epochMillis(diary.createdAt),
                updatedAtEpochMillis: Self.epochMillis(diary.updatedAt),
                photos: photos
            )
        }
        let envelope = DiaryBackupEnvelope(
            format: Self.formatIdentifier,
            version: Self.currentVersion,
            exportedAtEpochMillis: Self.epochMillis(now),
            diaries: entries
        )
        let encoder = JSONEncoder()
        encoder.outputFormatting = [.prettyPrinted, .sortedKeys]
        return try encoder.encode(envelope)
    }

    @discardableResult
    public func importData(_ data: Data) async throws -> Int {
        let envelope: DiaryBackupEnvelope
        do {
            envelope = try JSONDecoder().decode(DiaryBackupEnvelope.self, from: data)
        } catch {
            throw DiaryBackupError.invalidFormat
        }
        guard envelope.format == Self.formatIdentifier else {
            throw DiaryBackupError.invalidFormat
        }
        guard envelope.version == Self.currentVersion else {
            throw DiaryBackupError.unsupportedVersion
        }

        let existingById = Dictionary(
            uniqueKeysWithValues: try await repository.fetchAll().map { ($0.id, $0) }
        )
        var restoredCount = 0
        for entry in envelope.diaries {
            guard let id = UUID(uuidString: entry.id),
                  let mood = DiaryMood(rawValue: entry.mood),
                  entry.photos.count <= Diary.maximumPhotoCount else {
                throw DiaryBackupError.invalidFormat
            }

            let decodedPhotos = try entry.photos.map { photo -> Data in
                guard let photoData = Data(base64Encoded: photo.dataBase64),
                      photoData.count <= 10 * 1_024 * 1_024,
                      Self.sha256(photoData) == photo.sha256 else {
                    throw DiaryBackupError.invalidPhoto
                }
                return photoData
            }

            var newReferences: [String] = []
            do {
                for photoData in decodedPhotos {
                    newReferences.append(
                        try photoStore.saveJPEGData(photoData, diaryId: id)
                    )
                }
                let diary = Diary(
                    id: id,
                    userId: entry.userId,
                    title: entry.title,
                    body: entry.body,
                    mood: mood,
                    photoUrls: newReferences,
                    createdAt: Self.date(entry.createdAtEpochMillis),
                    updatedAt: Self.date(entry.updatedAtEpochMillis)
                )
                try await repository.save(diary)
                for oldReference in existingById[id]?.photoUrls ?? [] {
                    try? photoStore.delete(reference: oldReference)
                }
                restoredCount += 1
            } catch {
                for reference in newReferences {
                    try? photoStore.delete(reference: reference)
                }
                throw error
            }
        }
        return restoredCount
    }

    private static func sha256(_ data: Data) -> String {
        SHA256.hash(data: data).map { String(format: "%02x", $0) }.joined()
    }

    private static func epochMillis(_ date: Date) -> Int64 {
        Int64((date.timeIntervalSince1970 * 1_000).rounded())
    }

    private static func date(_ epochMillis: Int64) -> Date {
        Date(timeIntervalSince1970: TimeInterval(epochMillis) / 1_000)
    }
}
